nominal/core/datasource.py

```diff
@@ -0,0 +1,361 @@
+    storage_writer_api,
+    timeseries_channelmetadata,
+    timeseries_channelmetadata_api,
+    timeseries_logicalseries,
+    upload_api,
+)
+
+from nominal._utils import warn_on_deprecated_argument
+from nominal.core._clientsbunch import HasAuthHeader, ProtoWriteService
+from nominal.core._conjure_utils import _available_units, _build_unit_update
+from nominal.core._utils import HasRid
+from nominal.core.channel import Channel, ChannelDataType
+from nominal.ts import _MAX_TIMESTAMP, _MIN_TIMESTAMP, IntegralNanosecondsUTC, _SecondsNanos
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DataSource(HasRid):
+    rid: str
+    _clients: _Clients = field(repr=False)
+
+    class _Clients(Channel._Clients, HasAuthHeader, Protocol):
+        @property
+        def catalog(self) -> scout_catalog.CatalogService: ...
+        @property
+        def dataexport(self) -> scout_dataexport_api.DataExportService: ...
+        @property
+        def datasource(self) -> scout_datasource.DataSourceService: ...
+        @property
+        def logical_series(self) -> timeseries_logicalseries.LogicalSeriesService: ...
+        @property
+        def units(self) -> scout.UnitsService: ...
+        @property
+        def ingest(self) -> ingest_api.IngestService: ...
+        @property
+        def upload(self) -> upload_api.UploadService: ...
+        @property
+        def connection(self) -> scout_datasource_connection.ConnectionService: ...
+        @property
+        def storage_writer(self) -> storage_writer_api.NominalChannelWriterService: ...
+        @property
+        def proto_write(self) -> ProtoWriteService: ...
+        @property
+        def channel_metadata(self) -> timeseries_channelmetadata.ChannelMetadataService: ...
+
+    @warn_on_deprecated_argument(
+        "tags", "The 'tags' argument is deprecated because it's not used and will be removed in a future version."
+    )
+    def get_channel(self, name: str, tags: dict[str, str] | None = None) -> Channel:
+        for channel in self.get_channels(names=[name]):
+            if channel.name == name:
+                return channel
+        raise ValueError(f"channel {name!r} not found in dataset {self.rid!r}")
+
+    def get_channels(
+        self,
+        *,
+        names: Sequence[str] | None = None,
+    ) -> Iterable[Channel]:
+        """Look up the metadata for all matching channels associated with this datasource
+
+        Args:
+        ----
+            names: List of channel names to look up metadata for.
+
+        Yields:
+        ------
+            Yields a sequence of channel metadata objects which match the provided query parameters
+
+        """
+        if not names:
+            names = [channel.name for channel in self.search_channels()]
+
+        # Process in batches of 500
+        batch_size = 500
+        for i in range(0, len(names), batch_size):
+            batch_channel_names = names[i : i + batch_size]
```

Comment:
In general, you rarely need to actually slice with indexing. It's not a biggie really but it means that you need to require a sequence rather than a generic iterable. It's more flexible for consumers and for the future if you just pin yourself to _always_ allow a general iterable unless there's no other way around it.

Here I'd recommend copying the `itertools.batched` recommendation into a utils.py file - https://docs.python.org/3/library/itertools.html#itertools.batched

(copying because `batched` is only avail on 3.12+)